File: cathode/src/main/java/net/simonvt/cathode/remote/sync/shows/SyncShowsWatchlist.java
/*
 * Copyright (C) 2013 Simon Vig Therkildsen
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package net.simonvt.cathode.remote.sync.shows;

import android.database.Cursor;
import java.util.ArrayList;
import java.util.List;
import javax.inject.Inject;
import net.simonvt.cathode.api.entity.Show;
import net.simonvt.cathode.api.entity.WatchlistItem;
import net.simonvt.cathode.api.service.SyncService;
import net.simonvt.cathode.provider.DatabaseContract.ShowColumns;
import net.simonvt.cathode.provider.DatabaseSchematic;
import net.simonvt.cathode.provider.ProviderSchematic.Shows;
import net.simonvt.cathode.provider.ShowWrapper;
import net.simonvt.cathode.jobqueue.Job;

public class SyncShowsWatchlist extends Job {

  @Inject transient SyncService syncService;

  @Override public String key() {
    return "SyncShowsWatchlist";
  }

  @Override public int getPriority() {
    return PRIORITY_4;
  }

  @Override public void perform() {
    Cursor c = getContentResolver().query(Shows.SHOWS, new String[] {
        DatabaseSchematic.Tables.SHOWS + "." + ShowColumns.ID,
    }, ShowColumns.IN_WATCHLIST + "=0", null, null);

    List<Long> showIds = new ArrayList<Long>();

    while (c.moveToNext()) {
      showIds.add(c.getLong(c.getColumnIndex(ShowColumns.ID)));
    }
    c.close();

    List<WatchlistItem> watchlist = syncService.getShowWatchlist();

    for (WatchlistItem item : watchlist) {
      final Show show = item.getShow();
      final long listedAt = item.getListedAt().getTimeInMillis();
      final long traktId = show.getIds().getTrakt();

      long showId = ShowWrapper.getShowId(getContentResolver(), traktId);
      if (showId == -1L) {
        showId = ShowWrapper.createShow(getContentResolver(), traktId);
        queue(new SyncShow(traktId));
      }

      if (!showIds.remove(showId)) {
        ShowWrapper.setIsInWatchlist(getContentResolver(), showId, true, listedAt);
      }
    }

    for (Long showId : showIds) {
      ShowWrapper.setIsInWatchlist(getContentResolver(), showId, false);
    }
  }
}

File: cathode-sync/src/main/java/net/simonvt/cathode/actions/user/SyncShowsWatchlist.kt
/*
 * Copyright (C) 2013 Simon Vig Therkildsen
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package net.simonvt.cathode.actions.user

import android.content.Context
import androidx.work.WorkManager
import net.simonvt.cathode.actions.CallAction
import net.simonvt.cathode.actions.user.SyncShowsWatchlist.Params
import net.simonvt.cathode.api.entity.WatchlistItem
import net.simonvt.cathode.api.service.SyncService
import net.simonvt.cathode.common.database.forEach
import net.simonvt.cathode.common.database.getLong
import net.simonvt.cathode.provider.DatabaseContract.ShowColumns
import net.simonvt.cathode.provider.DatabaseSchematic
import net.simonvt.cathode.provider.ProviderSchematic.Shows
import net.simonvt.cathode.provider.helper.ShowDatabaseHelper
import net.simonvt.cathode.provider.query
import net.simonvt.cathode.settings.TraktTimestamps
import net.simonvt.cathode.work.enqueueUniqueNow
import net.simonvt.cathode.work.shows.SyncPendingShowsWorker
import retrofit2.Call
import javax.inject.Inject

class SyncShowsWatchlist @Inject constructor(
  private val context: Context,
  private val showHelper: ShowDatabaseHelper,
  private val syncService: SyncService,
  private val workManager: WorkManager
) : CallAction<Params, List<WatchlistItem>>() {

  override fun key(params: Params): String = "SyncShowsWatchlist"

  override fun getCall(params: Params): Call<List<WatchlistItem>> = syncService.getShowWatchlist()

  override suspend fun handleResponse(params: Params, response: List<WatchlistItem>) {
    val showIds = mutableListOf<Long>()

    val localWatchlist = context.contentResolver.query(
      Shows.SHOWS,
      arrayOf(DatabaseSchematic.Tables.SHOWS + "." + ShowColumns.ID),
      ShowColumns.IN_WATCHLIST
    )
    localWatchlist.forEach { cursor -> showIds.add(cursor.getLong(ShowColumns.ID)) }
    localWatchlist.close()

    for (watchlistItem in response) {
      val listedAt = watchlistItem.listed_at.timeInMillis
      val traktId = watchlistItem.show!!.ids.trakt!!
      val showResult = showHelper.getIdOrCreate(traktId)
      val showId = showResult.showId

      if (!showIds.remove(showId)) {
        showHelper.setIsInWatchlist(showId, true, listedAt)
      }
    }

    for (showId in showIds) {
      showHelper.setIsInWatchlist(showId, false)
    }

    workManager.enqueueUniqueNow(SyncPendingShowsWorker.TAG, SyncPendingShowsWorker::class.java)

    if (params.userActivityTime > 0L) {
      TraktTimestamps.getSettings(context)
        .edit()
        .putLong(TraktTimestamps.SHOW_WATCHLIST, params.userActivityTime)
        .apply()
    }
  }

  data class Params(val userActivityTime: Long = 0L)
}
